charts/seed-bootstrap/prometheus-rules/recording-rules.rules.yaml

```diff
@@ -36,6 +36,110 @@ groups:
   - record: seed:hvpa_status_applied_vpa_recommendation_target:sum_by_container
     expr: sum(hvpa_status_applied_vpa_recommendation{recommendation="target"}) by (container, resource)
 
+  # Recording rules for the count of containers which have CPU usage at or above 80% of their request, aggregated by containers.
+  - record: seed:containers_over_cpu_scaling_threshold_total:count_by_container
+    expr: count(max(rate(container_cpu_usage_seconds_total[5m])) by (container, namespace) / max(kube_pod_container_resource_requests_cpu_cores) by (container, namespace) >= 0.8) by (container)
+
+  # Recording rules for the count of containers which have memory usage at or above 120% of their request, aggregated by containers.
+  - record: seed:containers_over_memory_scaling_threshold_total:count_by_container
+    expr: count(max(container_memory_working_set_bytes) by (container, namespace) / max(kube_pod_container_resource_requests_memory_bytes) by (container, namespace) >= 1.2) by (container)
+
+  # Recording rules for the count of containers which have CPU usage at or above 80% of their limits, aggregated by containers.
+  - record: seed:containers_over_cpu_limiting_threshold_total:count_by_container
+    expr: count(max(rate(container_cpu_usage_seconds_total[5m])) by (container, namespace) / max(kube_pod_container_resource_limits_cpu_cores) by (container, namespace) >= 0.8) by (container)
+
+  # Recording rules for the count of containers which have memory usage at or above 80% of their limits, aggregated by containers.
+  - record: seed:containers_over_memory_limiting_threshold_total:count_by_container
+    expr: count(max(container_memory_working_set_bytes) by (container, namespace) / max(kube_pod_container_resource_limits_memory_bytes) by (container, namespace) >= 0.8) by (container)
+
+  # Recording rules for the count of containers which have CPU usage at or above 100% of their VPA recommendations, aggregated by containers.
```

Comment:
I don't think recording rules is the right approach for these metrics (or at least how it is implemented right now). What if we want to know the count of containers that have CPU usage at or above 80%? 50%? Or if we want to switch it around and know how many containers use less than 100%? To do this we would have to add the metrics and wait for a new gardener release. Also having this many recording rules is confusing and it is easy to lose oversight. Perhaps we can find a better way to do this.

Reply:
Agreed about the general goodness of using histograms for this. But I don't see any standard histogram metrics which I can use for this purpose. Please let me know if you are aware of any.

If you notice, all of the metrics below are at 100%. The rationale is to help tune the VPA and HPA to see it is recommending values right (safe enough and simultaneously not too generous). For this, 100% is sufficient and we do not need the intermediate percentiles.

In general, it would be desirable to create histogram metrics with quantile/percentiles. But that would be larger effort and will require more time. I would rather get these aggregate metrics in now and evaluate/tune VPA and HPA. We can work on histogram custom metrics later and replace these metrics. WDYT?

Reply:
What if we change the expression to:
```
max(rate(container_cpu_usage_seconds_total[5m])) by (container, namespace) / max(vpa_status_recommendation{recommendation="target", resource="cpu"}) by (container, namespace)
```
Basically, remove the comparison and the count calculation. It can be further granulised, but may not really be required as of now.
This will enable us to query all kinds of percentages and counts as mentioned by @wyb1 
Although, on the downside, this might increase the disk usage for prometheus, but I think we should be able to live with that.
Same thing can be done for lot of other recording rules here providing more flexibility, for eg, to figure out a favourable threshold percentage for HPA, we might need to compare against 130% also, and not just 80% or 120%.
WDYT?

Reply:
Nope. Cardinality = number of namespaces. Not good for aggregate prometheus.

Reply:
> We can work on histogram custom metrics later and replace these metrics. WDYT?

I understand you need these metrics to tune vpa/hvpa sooner rather than later so I am ok with this.

Howerver, I would at least like a metric that gives us the complement. So we know how many containers are using less than 100% of their given requests. This way we can account for all containers using VPA. Does this make sense?

Reply:
> Nope. Cardinality = number of namespaces. Not good for aggregate prometheus.

Just wanted to point out the benefits of added flexibility of comparisons by changing the expression (+ no need to roll out new release if we realise that we might need data for 130% threshold comparison for HPA for eg.).
After weighing pros and cons, if it still doesn't suit, then this works too 👍

Reply:
> Howerver, I would at least like a metric that gives us the complement. So we know how many containers are using less than 100% of their given requests. This way we can account for all containers using VPA. Does this make sense?

Yes. That makes sense. I will add the complementary metrics.

Reply:
> After weighing pros and cons, if it still doesn't suit, then this works too

The histogram approach would the right long term solution,

Reply:
@wyb1 I have added the complementary metrics.